Sources/ParseSwift/Coding/ParseEncoder.swift

```diff
@@ -0,0 +1,259 @@
+//
+
+// This rule doesn't allow types with underscores in their names.
+// swiftlint:disable type_name
+
+import Foundation
+
+// MARK: ParseEncoder
+public struct ParseEncoder {
+    let dateEncodingStrategy: AnyCodable.DateEncodingStrategy?
+    let jsonEncoder: JSONEncoder
+    let skippedKeys: Set<String>
+
+    init(
+        dateEncodingStrategy: AnyCodable.DateEncodingStrategy? = nil,
+        jsonEncoder: JSONEncoder = JSONEncoder(),
+        skippingKeys: Set<String> = []
+    ) {
+        self.dateEncodingStrategy = dateEncodingStrategy
+        self.jsonEncoder = jsonEncoder
+        self.skippedKeys = skippingKeys
+    }
+
+    func encodeToDictionary<T: Encodable>(_ value: T) throws -> [AnyHashable: Any] {
+        let encoder = _ParseEncoder(codingPath: [], dictionary: NSMutableDictionary(), skippingKeys: skippedKeys)
+        try value.encode(to: encoder)
+
+        // swiftlint:disable:next force_cast
+        return encoder.dictionary as! [AnyHashable: Any]
+    }
+
+    func encode<T: Encodable>(_ value: T) throws -> Data {
+        let dictionary = try encodeToDictionary(value)
+        return try jsonEncoder.encode(AnyCodable(dictionary, dateEncodingStrategy: dateEncodingStrategy!))
+    }
+
+    func encode<T: Encodable>(_ array: [T]) throws -> Data {
+        let dictionaries = try array.map { try encodeToDictionary($0) }
+        return try jsonEncoder.encode(AnyCodable(dictionaries, dateEncodingStrategy: dateEncodingStrategy!))
+    }
+}
+
+// MARK: _ParseEncoder
+internal struct _ParseEncoder: Encoder {
+    let codingPath: [CodingKey]
+    let dictionary: NSMutableDictionary
+    let skippedKeys: Set<String>
+    let userInfo: [CodingUserInfoKey: Any] = [:]
+
+    init(codingPath: [CodingKey], dictionary: NSMutableDictionary, skippingKeys: Set<String>) {
+        self.codingPath = codingPath
+        self.dictionary = dictionary
+        self.skippedKeys = skippingKeys
+    }
+
+    func container<Key>(keyedBy type: Key.Type) -> KeyedEncodingContainer<Key> where Key: CodingKey {
+        let container = _ParseEncoderKeyedEncodingContainer<Key>(
+            codingPath: codingPath,
+            dictionary: dictionary,
+            skippingKeys: skippedKeys
+        )
+
+        return KeyedEncodingContainer(container)
+    }
+
+    func singleValueContainer() -> SingleValueEncodingContainer {
+        _ParseEncoderSingleValueEncodingContainer(
+            codingPath: codingPath,
+            dictionary: dictionary,
+            skippingKeys: skippedKeys
+        )
+    }
+
+    func unkeyedContainer() -> UnkeyedEncodingContainer {
+        _ParseEncoderUnkeyedEncodingContainer(
+            codingPath: codingPath,
+            dictionary: dictionary,
+            skippingKeys: skippedKeys
+        )
+    }
+
+    static func encode<T: Encodable>(
+        _ value: T,
+        with codingPath: [CodingKey],
+        skippingKeys skippedKeys: Set<String>
+    ) throws -> Any {
+        switch value {
```

Comment:
Looks like my NSMutable speculation might be true?

Reply:
Can you clarify? I provided a comment about NS vs. the native Swift types above, but I'm not sure how it applies to this chunk of code.

Reply:
My comments here from earlier in reference to value for NSMutable, but I’m fine with your earlier response about the need for a reference and not making copies.